File: src/main/java/io/firetail/logging/util/UniqueIDGenerator.java
package io.firetail.logging.util;

import jakarta.servlet.http.HttpServletRequest;
import org.slf4j.MDC;

import java.util.UUID;

public class UniqueIDGenerator {

    private final String requestIdHeaderName;
    private final String correlationIdHeaderName;

    public UniqueIDGenerator(String requestIdHeaderName, String correlationIdHeaderName) {
        this.requestIdHeaderName = requestIdHeaderName;
        this.correlationIdHeaderName = correlationIdHeaderName;
    }

    public void generateAndSetMDC(HttpServletRequest request) {
        String requestId = request.getHeader(requestIdHeaderName);
        if (requestId == null)
            requestId = UUID.randomUUID().toString();
        MDC.put(requestIdHeaderName, requestId);

        String correlationId = request.getHeader(correlationIdHeaderName);
        if (correlationId == null)
            correlationId = UUID.randomUUID().toString();
        MDC.put(correlationIdHeaderName, correlationId);
    }

}

File: src/main/kotlin/io/firetail/logging/util/UniqueIDGenerator.kt
package io.firetail.logging.util

import io.firetail.logging.Constants.Companion.CORRELATION_ID
import io.firetail.logging.Constants.Companion.REQUEST_ID
import org.slf4j.MDC
import javax.servlet.http.HttpServletRequest

class UniqueIDGenerator(private val generator: Generator = Generator()) {
    fun generateAndSetMDC(request: HttpServletRequest) {
        MDC.clear()
        var requestId = request.getHeader(REQUEST_ID)
        if (requestId == null) requestId = generator.generate()
        MDC.put(REQUEST_ID, requestId)
        var correlationId = request.getHeader(CORRELATION_ID)
        if (correlationId == null) correlationId = generator.generate()
        MDC.put(CORRELATION_ID, correlationId)
    }
}
